File: frontend/web-client/src/app/services/branch.service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable, inject } from "@angular/core";
import { Observable } from "rxjs";
import { IBranch, IBranchRequest } from "../model/branch.model";
import { ILocationRequest } from "../model/location.model";
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: "root",
})
export class BranchService {
  private readonly _http = inject(HttpClient);
  private readonly _urlBase = `${environment.gatewayUrl}/api/v1/branches`;

  getBranches(): Observable<IBranch[]> {
    return this._http.get<IBranch[]>(this._urlBase);
  }

  getBranch(id: string): Observable<IBranch> {
    return this._http.get<IBranch>(`${this._urlBase}/${id}`);
  }

  getBranchesByCompany(companyId: string): Observable<IBranch[]> {
    return this._http.get<IBranch[]>(`${environment.gatewayUrl}/api/v1/companies/${companyId}/branches`);
  }

  createBranch(request: IBranchRequest): Observable<IBranch> {
    return this._http.post<IBranch>(this._urlBase, request);
  }

  createBranchForCompany(companyId: string, request: IBranchRequest): Observable<IBranch> {
    return this._http.post<IBranch>(`${environment.gatewayUrl}/api/v1/companies/${companyId}/branches`, request);
  }

  updateBranch(id: string, request: IBranchRequest): Observable<IBranch> {
    return this._http.patch<IBranch>(`${this._urlBase}/${id}`, request);
  }

  deleteBranch(id: string): Observable<void> {
    return this._http.delete<void>(`${this._urlBase}/${id}`);
  }

  createLocation(branchId: string, request: ILocationRequest): Observable<IBranch> {
    return this._http.post<IBranch>(`${this._urlBase}/${branchId}/locations`, request);
  }

  updateLocation(branchId: string, locationId: string, request: ILocationRequest): Observable<IBranch> {
    return this._http.patch<IBranch>(`${this._urlBase}/${branchId}/locations/${locationId}`, request);
  }

  deleteLocation(branchId: string, locationId: string): Observable<IBranch> {
    return this._http.delete<IBranch>(`${this._urlBase}/${branchId}/locations/${locationId}`);
  }
}

File: backend/erphub-api/microservices/branch-sv/src/main/java/com/crishof/branchsv/service/BranchService.java
package com.crishof.branchsv.service;

import com.crishof.branchsv.dto.BranchRequest;
import com.crishof.branchsv.dto.BranchResponse;
import com.crishof.branchsv.dto.LocationRequest;
import com.crishof.branchsv.exception.BranchNotFoundException;
import com.crishof.branchsv.exception.CompanyNotFoundException;
import com.crishof.branchsv.model.Branch;

import java.util.List;
import java.util.UUID;

public interface BranchService {

    List<BranchResponse> getAllBranches();

    List<BranchResponse> getBranchesByCompanyId(UUID companyId);

    BranchResponse getBranchResponseById(UUID id) throws BranchNotFoundException;

    Branch getBranchById(UUID branchId) throws BranchNotFoundException;

    BranchResponse createBranch(BranchRequest branchRequest) throws CompanyNotFoundException;

    BranchResponse createBranchForCompany(UUID companyId, BranchRequest branchRequest) throws CompanyNotFoundException;

    BranchResponse updateBranch(UUID branchId, BranchRequest branchRequest) throws BranchNotFoundException, CompanyNotFoundException;

    void deleteBranch(UUID branchId) throws BranchNotFoundException;

    BranchResponse createLocation(UUID branchId, LocationRequest request);

    BranchResponse updateLocation(UUID branchId, UUID locationId, LocationRequest request) throws BranchNotFoundException;

    BranchResponse deleteLocation(UUID branchId, UUID locationId) throws BranchNotFoundException;
}

